src/x11.rs

```diff
@@ -74,19 +105,96 @@ impl X11Impl {
         let visual = (lib.XDefaultVisual)(display_handle.display as *mut Display, screen);
         let depth = (lib.XDefaultDepth)(display_handle.display as *mut Display, screen);
 
+        // See if we can load the XShm extension.
+        let xshm = XShm::open()
+            .ok()
+            .filter(|shm| (shm.XShmQueryExtension)(display_handle.display as *mut Display) == 0);
+
         Ok(Self {
             window_handle,
             display_handle,
-            lib,
+            xlib: lib,
+            xshm: xshm.map(|xshm| ShmExtension { xshm, shmseg: None }),
             gc,
             visual,
             depth,
         })
     }
 
     pub(crate) unsafe fn set_buffer(&mut self, buffer: &[u32], width: u16, height: u16) {
+        if let Err(e) = self.shm_set(buffer, width, height) {
+            eprintln!("XShm not available, falling back to XImage: {e}");
+            self.fallback_set(buffer, width, height);
+        }
+    }
+
+    /// Set the buffer to the given image using shared memory.
+    unsafe fn shm_set(&mut self, buffer: &[u32], width: u16, height: u16) -> io::Result<()> {
+        let shm_ext = match self.xshm.as_mut() {
+            Some(shm_ext) => shm_ext,
+            None => return Err(io::Error::new(io::ErrorKind::Other, "XShm not available")),
+        };
+
+        // Get the size of the shared memory segment.
+        let shmseg_size = (width as usize)
+            .checked_mul(height as usize)
+            .and_then(|size| size.checked_mul(4))
+            .expect("Buffer size overflow");
+
+        // Create the shared memory segment if it doesn't exist, or if it's the wrong size.
+        let shmseg = match &mut shm_ext.shmseg {
+            None => shm_ext.shmseg.insert(ShmSegment::new(shmseg_size)?),
+            Some(ref shmseg) if shmseg.size < shmseg_size => {
+                shm_ext.shmseg.insert(ShmSegment::new(shmseg_size)?)
+            }
+            Some(shmseg) => shmseg,
+        };
+
+        // Create the basic image.
+        let mut seg: XShmSegmentInfo = mem::zeroed();
```

Comment:
I'd think using MaybeUninit would be clearer

Reply:
I don't know if `mem::zeroed()` is unclear. Seems fine to meas long as all zeros is a valid bit pattern for this type (otherwise it would be unsound).

Using uninitialized memory could technically be faster, and still sound, but it may be safer to stick with either `mem::zeroed()` or `MaybeUninit::zeroed()`.

Reply:
Once #52 is merged I'll rewrite this PR to use `x11rb`, which should circumvent this issue entirely.